feat: Submission flow redesign

olerichter00 opened the pull request:
# This PR is ready to be reviewed ✅

This PR resolves [ONYX-908] <!-- eg [PROJECT-XXXX] -->

### Description

This PR adds the backbone of the upcoming phase 1 work on the new submission flow.

What this PR does:
- Add the main navigation logic
- Add a couple of required screens
- Set up test helpers and add some tests
- Store and schema validation logic


https://github.com/artsy/eigen/assets/11945712/98eac8ba-7ea5-4289-a8e5-1fb7dd0fe3e1

The remaining work has been added to the epic https://artsyproduct.atlassian.net/browse/ONYX-951


### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->
<!-- ⚠️ Prefix with `[NEEDS EXTERNAL QA]` if a change requires external QA -->

#### Cross-platform user-facing changes

- add new submission flow screens - mounir

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[ONYX-908]: https://artsyproduct.atlassian.net/browse/ONYX-908?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ArtsyOpenSource commented:
<!--
  0 failure: 
  1 warning:  ❌ **An error occu...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  ❌ **An error occurred while validating your changelog, please make sure you provided a valid changelog.**
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 98e3604ac52060ea2fc24717c1045141511cbe91
</p>

olerichter00 commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Utils/validation.ts:
```diff
@@ -0,0 +1,214 @@
+  userPhone: "",
+}
+
+// export const artworkDetailsValidationSchema = Yup.object().shape({
```
This can be removed 👀

olerichter00 reviewed (commented):
Looks amazing! 🌟

nickskalkin commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkSelectArtist.tsx:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    try {
+      // TODO: Does it make sense to create a new submission here?
```
I don't see a problem with it + we can cleanup abandoned submissions on backend

MounirDhahri replied:
I can bring this up to the next product update to discuss it together

MounirDhahri commented on src/app/AppRegistry.tsx:
```diff
@@ -463,7 +464,7 @@ export const modules = defineModules({
   ConversationDetails: reactModule(ConversationDetailsQueryRenderer),
   DarkModeSettings: reactModule(DarkModeSettings),
-  DevMenu: reactModule(DevMenu, { fullBleed: true, hidesBottomTabs: true, hidesBackButton: true }),
+  DevMenu: reactModule(DevMenu, { hidesBottomTabs: true, hidesBackButton: true }),
```
I needed this because it allowed me to avoid handling the safe area inside the modal and instead get that directly thanks to this.

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/ArtistAutosuggest.tsx:
```diff
@@ -12,16 +12,20 @@ import { IMAGE_SIZE } from "app/Scenes/Search/components/SearchResultImage"
 export const ArtistAutosuggest: React.FC<ArtistAutosuggestProps> = ({
   onResultPress,
   onSkipPress,
+  disableCustomArtistCreation,
```
good idea - will change this

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/ArtistAutosuggest.tsx:
```diff
@@ -36,9 +40,17 @@ export const ArtistAutosuggest: React.FC<ArtistAutosuggestProps> = ({
-  const collectedArtists = extractNodes(queryData.me?.userInterestsConnection).filter(
-    (node) => node.__typename === "Artist"
-  )
+  const collectedArtists = extractNodes(queryData.me?.userInterestsConnection).filter((node) => {
+    if (node.__typename === "Artist") {
+      if (onlyP1Artists) {
+        return node.targetSupply?.priority === "TRUE"
+      }
+      return true
+    }
+
+    return false
+  })
```
much more readable - will update

MounirDhahri commented on src/app/Scenes/MyCollection/Screens/ArtworkForm/Components/ArtistAutosuggest.tsx:
```diff
@@ -169,6 +188,7 @@ const ArtistAutosuggestScreenQuery = graphql`
               }
               targetSupply {
                 isP1
+                priority
```
yes, updated this in the code and forgot to remove it here. Will do it now

nickskalkin replied:
I am wondering if instead of using `priority` field, which returns "TRUE" or "FALSE" as enum (string?), we can use `isTargetSupply` ([code](https://github.com/artsy/metaphysics/blob/main/src/schema/v2/artist/targetSupply/index.ts#L53)). If I understand it correctly, it is backed by the same gravity field (see [here](https://github.com/artsy/gravity/blob/main/app/models/domain/artist.rb#L576) - gravity uses `target_supply_priority` to calculate this field), but returns boolean true / false - definitely more convenient to work with it.

cc @MounirDhahri @olerichter00

MounirDhahri commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkAddDetails.tsx:
```diff
@@ -0,0 +1,171 @@
+export const SubmitArtworkAddDetails = () => {
+  const [isRarityInfoModalVisible, setIsRarityInfoModalVisible] = useState(false)
+  const [isYearUnknown, setIsYearUnknown] = useState(false)
+  const [oldTypedYear, setOldTypedYear] = useState("")
```
I actually thought about that but since the `createOrUpdate` mutation helper we had was quite sensitive to wrong input fields - I figured out it's okay to keep this here and decouple it from the submit logic. Glad to change approach if you strongly feel this should not be here.

olerichter00 replied:
Let's leave it here, probably good to have the logic within this component 👍 

Probably a good idea to, at some point, specify all values passed to the mutation explicitly and not take them directly from the form to avoid this.

MounirDhahri commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkAddDimensions.tsx:
```diff
@@ -0,0 +1,115 @@
+    [dimensionMetric]
+  )
+
+  // Doing some calculations to determine a good width for the input fields to make sure
+  // it doesn't look too cramped in small devices
+  const isBigScreen = screenWidth >= 375
+  const inputWidth = isBigScreen ? 110 : screenWidth / 2 - 2 * space(2)
```
I will bring this back to how it was then and discuss it with him before making this decision on my own

MounirDhahri commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkCompleteYourSubmission.tsx:
```diff
@@ -0,0 +1,54 @@
+
+      <Message
+        title="Next steps"
+        // TODO: Add this support to the palette Message component
+        // @ts-expect-error
```
I am on it now 👍

MounirDhahri commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkSelectArtist.tsx:
```diff
@@ -0,0 +1,99 @@
+  const formik = useFormikContext<ArtworkDetailsFormModel>()
+
+  useEffect(() => {
+    fetchUserContactInformation()
```
Valid point

MounirDhahri replied:
oh yeah - just remembered why I moved there 😄 I wrote it first in the file wrapped with the store then moved it here because I struggled with injecting the value initially. I will look into it again

olerichter00 replied:
Oh, it might be more difficult with Formik. If this isn't easy peasy, let's just leave the logic here or maybe move it to `SubmitArtworkFormContent`.

MounirDhahri replied:
my concern didn't make sense - it worked. Glad to be proven wrong

MounirDhahri commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkStartFlow.tsx:
```diff
@@ -0,0 +1,61 @@
+
+const SECTIONS = [
+  {
+    title: "Add your artwork lorem ipsum",
```
I intentionally left this with lorem ipsum because the copy here is still not ready unfortunately.

MounirDhahri commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkSelectArtist.tsx:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    try {
+      // TODO: Does it make sense to create a new submission here?
+      // We might end up with a lot of submissions that are never completed
```
Good idea

MounirDhahri commented:
Whaaaat whyyyy

MounirDhahri commented:
@olerichter00 addressed all your comments - let me know if I need to make any other changes

olerichter00 commented:
> @olerichter00 addressed all your comments - let me know if I need to make any other changes

That was fast! I've looked through the code again, and everything looks great to me! 🌟

nickskalkin commented on src/app/Scenes/SellWithArtsy/ArtworkForm/Components/SubmitArtworkAddDetails.tests.tsx:
```diff
@@ -0,0 +1,73 @@
+    const mediumPicker = screen.getByTestId("CategorySelect")
+    expect(mediumPicker).toBeOnTheScreen()
+    fireEvent.press(mediumPicker)
+    // Wait for the select modal to show up
```
big thanks for comments before `flushPromiseQueue` - it can be very confusing why it is there without explanation.

MounirDhahri replied:
you're welcome ❤️

MounirDhahri commented:
Going to merge this to allow us to test things out on the old flow further
